### PR TITLE
[#554] Add WalletConfigProvider

### PR DESCRIPTION
**Pull request by Chlup**
- Added `FeaturesFlagManager` object which handles loading and caching of feature flags.
- Added `FeatureFlagsConfiguration` which represents one configuration of all feature flags.

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [x] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+typealias UserDefaultsFeatureFlagsConfigurationProvider = UserDefaultsFeatureFlagsStorage
+typealias UserDefaultsFeatureFlagsManagerCache = UserDefaultsFeatureFlagsStorage
+
+struct UserDefaultsFeatureFlagsStorage {
+    private let userDefaults = UserDefaults.standard
+
+    enum InternalError: Error {
+        case nothingStored
+        case cantDeserializeData
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:18]:
```suggestion
        case unableToDeserializeData
```

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import Foundation
+
+typealias UserDefaultsFeatureFlagsConfigurationProvider = UserDefaultsFeatureFlagsStorage
+typealias UserDefaultsFeatureFlagsManagerCache = UserDefaultsFeatureFlagsStorage
+
+struct UserDefaultsFeatureFlagsStorage {
+    private let userDefaults = UserDefaults.standard
+
+    enum InternalError: Error {
+        case nothingStored
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:17]:
```suggestion
        case noValueStored
```

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        static let cacheKey = "feature_flags_ud_config_cache"
+    }
+
+    private func load(key: String) async throws -> FeatureFlagsConfiguration {
+        guard let data = userDefaults.data(forKey: key) else { throw InternalError.nothingStored }
+        do {
+            let rawFlags = try PropertyListDecoder().decode(FeatureFlagsConfiguration.RawFlags.self, from: data)
+            return FeatureFlagsConfiguration(flags: rawFlags)
+        } catch {
+            LoggerProxy.debug("Error when deocding feature flags from user defaults: \(error)")
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:32]:
```suggestion
            LoggerProxy.debug("Error when decoding feature flags from user defaults: \(error)")
```

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            throw InternalError.cantDeserializeData
+        }
+    }
+
+    private func store(flags: FeatureFlagsConfiguration.RawFlags, key: String) async {
+        do {
+            let data = try PropertyListEncoder().encode(flags)
+            userDefaults.set(data, forKey: key)
+        } catch {
+            LoggerProxy.debug("Can't store/encode feature flags when updating user defaults: \(error)")
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:42]: should this be an actor then?

File: secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift [hunk 5]
```diff
@@ -0,0 +1,78 @@
+//
+//  FeatureFlagsManager.swift
+//  secant
+//
+//  Created by Michal Fousek on 23.02.2023.
+//
+
+import Foundation
+
+struct FeatureFlagsManager {
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift:10]:
```suggestion
struct ConfigurationManager {
```

File: secant/Dependencies/FeatureFlagsManager/FeatureFlagsManagerInterface.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+//  secant
+//
+//  Created by Michal Fousek on 23.02.2023.
+//
+
+import ComposableArchitecture
+import Foundation
+
+extension DependencyValues {
+    var featureFlagsManager: FeatureFlagsManagerClient {
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/FeatureFlagsManagerInterface.swift:12]:
```suggestion
    var configurationManager: ConfigurationManagerClient {
```

File: secant/Dependencies/FeatureFlagsManager/FeatureFlagsManagerInterface.swift [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        get { self[FeatureFlagsManagerClient.self] }
+        set { self[FeatureFlagsManagerClient.self] = newValue }
+    }
+}
+
+struct FeatureFlagsManagerClient {
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/FeatureFlagsManagerInterface.swift:18]:
```suggestion
struct ConfigurationManagerClient {
```

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 8]
```diff
@@ -0,0 +1,75 @@
+//
+//  UserDefaultsFeatureFlagsStorage.swift
+//  secant
+//
+//  Created by Michal Fousek on 23.02.2023.
+//
+
+import Foundation
+
+typealias UserDefaultsFeatureFlagsConfigurationProvider = UserDefaultsFeatureFlagsStorage
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:10]:
```suggestion
typealias UserDefaultsConfigurationProvider = UserDefaultsConfigurationStorage
```

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+typealias UserDefaultsFeatureFlagsManagerCache = UserDefaultsFeatureFlagsStorage
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:11]:
```suggestion
typealias UserDefaultsConfigurationManagerCache = UserDefaultsFeatureFlagsStorage
```

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+struct UserDefaultsFeatureFlagsStorage {
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:13]:
```suggestion
struct UserDefaultsConfigurationStorage {
```

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    enum Constants {
+        static let providerKey = "feature_flags_ud_config_provider"
+        static let cacheKey = "feature_flags_ud_config_cache"
+    }
+
+    private func load(key: String) async throws -> FeatureFlagsConfiguration {
+        guard let data = userDefaults.data(forKey: key) else { throw InternalError.nothingStored }
+        do {
+            let rawFlags = try PropertyListDecoder().decode(FeatureFlagsConfiguration.RawFlags.self, from: data)
+            return FeatureFlagsConfiguration(flags: rawFlags)
```
Comment (pacu) [inline secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift:30]:
```suggestion
            return Configuration(flags: rawFlags)
```

File: secant/Models/FeatureFlags.swift [hunk 12]
```diff
@@ -0,0 +1,34 @@
+//
+//  FeatureFlags.swift
+//  secant
+//
+//  Created by Michal Fousek on 23.02.2023.
+//
+
+enum FeatureFlag: String, CaseIterable, Codable {
```
Comment (pacu) [inline secant/Models/FeatureFlags.swift:8]:
```suggestion
enum Setting: String, CaseIterable, Codable {
```

File: secant/Models/FeatureFlags.swift [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+    var enabledByDefault: Bool {
+        switch self {
+        case .testFlag1, .testFlag2:
+            return false
+        }
+    }
+}
+
+struct FeatureFlagsConfiguration: Equatable {
```
Comment (pacu) [inline secant/Models/FeatureFlags.swift:21]:
```suggestion
struct Configuration: Equatable {
```

**Review by pacu — CHANGES_REQUESTED** [review]
I left some comments and suggestions on naming. If you figure out better names they are totally welcome! I think that the settings we will end up storing are more than just feature flags so that name won't be adequate mid term

File: secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    /// Objects that fetches flags configuration from some source. It can be fetched from user defaults or some backend API for example. It depends
+    /// on implementation.
```
[the remark this answers is not in the document]
Comment (Chlup) [reply secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift]: We have 150 chars limit on line setup in linter. Whole comment simply doesn't fit one line. 

Screenshot 2023-02-24 at 10 16 41

File: secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (Chlup) [reply secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift]: I renamed this to `WalletConfigProvider`.

File: secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+        let finalConfiguration = merge(configuration: configuration, withDefaultConfiguration: FeatureFlagsConfiguration.default)
+
+        await cache.store(finalConfiguration)
+
+        return finalConfiguration
+    }
+
+    // This is used only in debug menu to change configuration for specific flag
+    func update(featureFlag: FeatureFlag, isEnabled: Bool) async {
```
[the remark this answers is not in the document]
Comment (Chlup) [reply secant/Dependencies/FeatureFlagsManager/FeatureFlagsManager.swift]: Purpose of this method for now is to update value for feature flag. It's this specific purpose. This method will be used only from debug screen. Maybe once we have some server where we can setup flags and other configuration we can drop this function. 

It has specific function and therefore it has specific name.

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift (hunk 9 above)
Comment (Chlup) [reply secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift]: I renamed this to `WalletConfigProviderCache`.

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift (hunk 10 above)
Comment (Chlup) [reply secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift]: I renamed this to `UserDefaultsWalletConfigStorage`.

File: secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift (hunk 11 above)
Comment (Chlup) [reply secant/Dependencies/FeatureFlagsManager/UserDefaultsFeatureFlagsStorage.swift]: I renamed `FeatureFlagsConfiguration` to `WalletConfig`.

File: secant/Models/FeatureFlags.swift (hunk 12 above)
Comment (Chlup) [reply secant/Models/FeatureFlags.swift]: This enum has one specific purpose. And that is to hold list of feature flags. In future we may have some other things (AB tests, server side UI, whatever....). And data structures representing these things will probably have specific purposes and names. For example:
```
enum ABTestName: String {
....
}
```

Then all the specific things will be inside `WalletConfig` structure. `Settings` seems like alternative name for `WalletConfig`. But it seems to generic for enum that holds list of feature flags.

File: secant/Models/FeatureFlags.swift (hunk 13 above)
Comment (Chlup) [reply secant/Models/FeatureFlags.swift]: I renamed this to `WalletConfig`

**Review by pacu — APPROVED** [review]
utACK!

Please change the folder name of Dependencies/FeatureFlags... to the new name before merging

**Pull request merged**